packages/engine/paima-funnel/src/cde/cardanoProjectedNFT.ts

```diff
@@ -0,0 +1,67 @@
+import type {
+  CdeCardanoProjectedNFTDatum,
+  ChainDataExtensionCardanoProjectedNFT,
+  ChainDataExtensionDatum,
+} from '@paima/sm';
+import { ChainDataExtensionDatumType, DEFAULT_FUNNEL_TIMEOUT, timeout } from '@paima/utils';
+import { Routes, query } from '@dcspark/carp-client/client/src';
+import type { ProjectedNftRangeResponse } from '@dcspark/carp-client/shared/models/ProjectedNftRange';
+
+export default async function getCdeProjectedNFTData(
+  url: string,
+  extension: ChainDataExtensionCardanoProjectedNFT,
+  fromAbsoluteSlot: number,
+  toAbsoluteSlot: number,
+  getBlockNumber: (slot: number) => number
+): Promise<ChainDataExtensionDatum[]> {
+  const events = await timeout(
+    query(url, Routes.projectedNftEventsRange, {
+      range: { minSlot: fromAbsoluteSlot, maxSlot: toAbsoluteSlot },
+    }),
+    DEFAULT_FUNNEL_TIMEOUT
+  );
+
+  return events
+    .map(e => eventToCdeDatum(e, extension, getBlockNumber(e.actionSlot)))
+    .filter(e => e != null)
+    .map(e => e!);
```

Comment:
why?

Reply:
@ecioppettini `e!` is special typescript syntax to ignore the nullability of a variable. Generally for cases like this I prefer to use reduce since it gives better type inference (so you don't have to use `e!`)